### DISCUSSION TITLE
BPX v0.6.0 candidate

### DISCUSSION THREAD

**Discussion by rtimms** [FaradayInstitution/BPX#62] [Ideas]
Opening this to discuss candidate features/issues to tackle for BPX v0.6.0.

Below are links to some issues that are already open and could form the basis of v0.6.0, but it would also be good to discuss "bigger picture" thoughts on where BPX needs to go.

## Features (modelling)
- Add support for degradation modes https://github.com/FaradayInstitution/BPX/issues/60
- Explicit support for hysteresis 
- More model support, e.g. specify SPMe (see https://github.com/FaradayInstitution/BPX/issues/59) or MSMR
- More flexible "User-defined" section https://github.com/FaradayInstitution/BPX/issues/46
- Extended operations in BPX functions https://github.com/FaradayInstitution/BPX/issues/53
- More examples of validation data https://github.com/FaradayInstitution/BPX/issues/61

## Features (JSON structure/metadata)
- Add version number to BPX file for "evolving" parameter sets https://github.com/FaradayInstitution/BPX/issues/25
- Add simulation metadata to assist with validation https://github.com/FaradayInstitution/BPX/issues/28 (also see https://github.com/FaradayInstitution/BPX/discussions/22#discussioncomment-6250116)
- Allow partial BPX files https://github.com/FaradayInstitution/BPX/issues/40
- Revise hierarchy of "Electrode" parameters https://github.com/FaradayInstitution/BPX/issues/55

## Bug fixes
- Irrelevant validation errors https://github.com/FaradayInstitution/BPX/issues/52

**Comment by rtimms** [comment]
Simplify specification of active material content #41

**Comment by ejfdickinson** [comment]
Added #66 as a small issue

**Comment by jsimonclark** [comment]
added #68 as enhancement for JSON structure/metadata

**Comment by rtimms** [comment]
This will be v0.6.0 now. We made a v0.5.0 which has the same functionality as 0.4.0 but updates to use pydantic v2 (support for v1 was dropped)

**Comment by ikorotkin** [comment]
From DandeLiion's point of view, the most wanted feature is to allow the "User-defined" section to be *anything* (a valid JSON in general), including entire JSON branches to specify new models not supported by BPX. Related to #46 and #67.
We currently have to use various tricks to work around this restriction. BPX cannot facilitate all possible solvers and (sub)models. A permissive user-defined section would be very useful.

Other issues that in my opinion would be worth looking into:

- The thermal parameters are inconsistent in BPX schema (related issue #89)
- Initial conditions are defined in the schema, but they are not complete (related issue #4).
    - Initial temperature is in the `"Cell"` section, `"Initial temperature [K]"` field
    - Initial concentration in electrolyte is in the `"Electrolyte"` section, `"Initial concentration in electrolyte [mol.m-3]"` field
    - That is not enough to start a simulation, initial state of charge or initial voltage should be defined.
    - If BPX does not need to define the initial conditions, `"Initial temperature [K]"` should be removed as well and `"Initial concentration in electrolyte [mol.m-3]"` renamed to "Typical" or "Equilibrium" concentration in electrolyte. Otherwise, initial SoC and voltage should be defined (perhaps, in a new section related to the initial state?)
- `FloatFunctionTable` does not accept `int` values (issue #66)

**Reply by ejfdickinson** [reply]
My suggestion for solving #4 is that we split all the "environmental" or "state" parameters out separately, to include all types of boundary conditions (initial values and couplings, e.g. heat transfer coefficient for lumped thermal model as in #89 ). This section becomes optional, but specified, and we try to make sure it's complete.

**Reply by ikorotkin** [reply]
That's a great idea, thanks @ejfdickinson!

The `"Ambient temperature [K]"` should be optional and go to the "environmental" kind of section as well (currently, it's a required parameter in the schema, unlike other thermal parameters, as mentioned in the comments in #89).
...And the `"Thermal conductivity [W.m-1.K-1]"` should be removed as it is not even used anywhere in the BPX doc (#50).
